### PR TITLE
Add full case conversions

### PR DESCRIPTION
**Pull request by wismill**
Add full case conversions as in `text`. It does not cover context- nor locale-dependent conversions.

See #47.

### CODE REVIEW

File: unicode-data/exe/Parser/Text.hs [hunk 1]
```diff
@@ -598,6 +599,116 @@ genSimpleCaseMappingModule moduleName funcName field =
...
+
+    mkSimple = fmap ord . simple
+    mkSpecial = fmap (encode . special) . (specialCasings Map.!?) . _char
+    encode :: String -> Int
+    encode
+        = foldr (\(k, c) -> (+) (ord c `shiftL` k)) 0
+        . zip [0, 21, 42]
+        -- Check min 1 character, max 3 characters
+        . (\cs -> if null cs || length cs > 3 then error (show cs) else cs)
+
```
Comment (adithyaov) [inline unicode-data/exe/Parser/Text.hs:660]:
```suggestion
```

**Review by adithyaov — APPROVED** [review]
Again, Apologies for the extremely delayed review.

**Comment by wismill** [comment]
@harendra-kumar I will merge this tomorrow if there is no further comment.

File: unicode-data/lib/Unicode/Char/Case.hs [hunk 2]
```diff
@@ -47,3 +77,153 @@ isUpperCase = P.isUppercase
...
+-- | Extract the next character from a raw case mapping.
+--
+-- **Note:** this does not work for character @\'\\NUL\'@
+{-# INLINE step #-}
+step :: Int64 -> Step Int64 Char
+step = \case
+    0 -> Stop
+    s -> Yield (chr cp) (s `shiftR` 21)
+        where
+            -- (1 << 21) - 1: bits used for a Unicode code point.
```
Comment (harendra-kumar) [inline unicode-data/lib/Unicode/Char/Case.hs:227]: Why not just say 21 bits?

File: unicode-data/lib/Unicode/Char/Case.hs [hunk 3]
```diff
@@ -47,3 +77,153 @@ isUpperCase = P.isUppercase
...
+        k -> k
+
+-- [TODO] @since
+-- | Returns the full /upper/ case mapping of a character.
+--
+-- It uses the character property @Uppercase_Mapping@.
+--
+-- __Note:__ @\'\\NUL\'@ requires special handling.
+-- See 'toUpperString' source code for an example.
+upperCaseMapping :: Unfold Char Char
```
Comment (harendra-kumar) [inline unicode-data/lib/Unicode/Char/Case.hs:140]: Add INLINE.

File: unicode-data/lib/Unicode/Char/Case.hs [hunk 4]
```diff
@@ -47,3 +77,153 @@ isUpperCase = P.isUppercase
...
+        k -> k
+
+-- [TODO] @since
+-- | Returns the full /title/ case mapping of a character.
+--
+-- It uses the character property @Titlecase_Mapping@.
+--
+-- __Note:__ @\'\\NUL\'@ requires special handling.
+-- See 'toTitleString' source code for an example.
+titleCaseMapping :: Unfold Char Char
```
Comment (harendra-kumar) [inline unicode-data/lib/Unicode/Char/Case.hs:126]: Add INLINE.

File: unicode-data/lib/Unicode/Char/Case.hs [hunk 5]
```diff
@@ -47,3 +77,153 @@ isUpperCase = P.isUppercase
...
+        k -> k
+
+-- [TODO] @since
+-- | Returns the full /lower/ case mapping of a character.
+--
+-- It uses the character property @Lowercase_Mapping@.
+--
+-- __Note:__ @\'\\NUL\'@ requires special handling.
+-- See 'toLowerString' source code for an example.
+lowerCaseMapping :: Unfold Char Char
```
Comment (harendra-kumar) [inline unicode-data/lib/Unicode/Char/Case.hs:112]: Add INLINE.

File: unicode-data/lib/Unicode/Char/Case.hs [hunk 6]
```diff
@@ -47,3 +77,153 @@ isUpperCase = P.isUppercase
...
+-- module are /not/ locale nor context sensitive.
+
+-- [TODO] @since
+-- | Returns the full /folded/ case mapping of a character.
+--
+-- It uses the character property @Case_Folding@.
+--
+-- __Note:__ @\'\\NUL\'@ requires special handling.
+-- See 'toCaseFoldString' source code for an example.
+caseFoldingMapping :: Unfold Char Char
```
Comment (harendra-kumar) [inline unicode-data/lib/Unicode/Char/Case.hs:98]: Add INLINE.

How about renaming to `caseFoldMapping` to keep it consistent with other names like `lowerCaseMapping`?

**Review by harendra-kumar — CHANGES_REQUESTED** [review]
Sorry, I just missed this PR. The change looks pretty good. I have a few minor suggestions.

File: unicode-data/lib/Unicode/Internal/Unfold.hs [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+    -- ^ Produces a single value and the next state of the stream.
+    | Stop
+    -- ^ Indicates there are no more values in the stream.
+
+instance Functor (Step s) where
+    {-# INLINE fmap #-}
+    fmap f (Yield x s) = Yield (f x) s
+    fmap _ Stop = Stop
+
+toList :: Unfold a b -> a -> [b]
```
[the remark this answers is not in the document]
Comment (wismill) [reply unicode-data/lib/Unicode/Internal/Unfold.hs:43]: Done

File: unicode-data/lib/Unicode/Char/Case.hs (hunk 3 above)
Comment (wismill) [reply unicode-data/lib/Unicode/Char/Case.hs:144]: Done

File: unicode-data/lib/Unicode/Char/Case.hs (hunk 4 above)
Comment (wismill) [reply unicode-data/lib/Unicode/Char/Case.hs:129]: Done

File: unicode-data/lib/Unicode/Char/Case.hs (hunk 6 above)
Comment (wismill) [reply unicode-data/lib/Unicode/Char/Case.hs]: Done

File: unicode-data/lib/Unicode/Char/Case.hs (hunk 5 above)
Comment (wismill) [reply unicode-data/lib/Unicode/Char/Case.hs:114]: Done

File: unicode-data/lib/Unicode/Char/Case.hs (hunk 2 above)
Comment (wismill) [reply unicode-data/lib/Unicode/Char/Case.hs]: Done

**Pull request merged**